core/domain/stats_services.py

```diff
@@ -18,23 +18,166 @@
 
 import itertools
 
-from core.domain import exp_services
+from core.domain import exp_domain
 from core.domain import interaction_registry
 from core.domain import stats_domain
-from core.domain import stats_jobs_continuous
 from core.platform import models
 
 (stats_models,) = models.Registry.import_models([models.NAMES.statistics])
 
+# Counts contributions from all versions.
+VERSION_ALL = 'all'
+
+def handle_stats_creation_for_new_exploration(exp_id, exp_version, state_names):
+    """Creates ExplorationStatsModel for the freshly created exploration and
+    sets all initial values to zero.
+
+    Args:
+        exp_id: str. ID of the exploration.
+        exp_version. int. Version of the exploration.
+        state_names: list(str). State names of the exploration.
+    """
+    state_stats_mapping = {
+        state_name: stats_domain.StateStats.create_default()
+        for state_name in state_names
+    }
+
+    exploration_stats = stats_domain.ExplorationStats(
+        exp_id, exp_version, 0, 0, state_stats_mapping)
+    create_stats_model(exploration_stats)
+
+
+def handle_stats_creation_for_new_exp_version(exploration, change_list):
+    """Retrieves the ExplorationStatsModel for the old exp_version and makes
+    any required changes to the structure of the model. Then, a new
+    ExplorationStatsModel is created for the new exp_version.
+
+    Args:
+        exploration: Exploration. The exploration domain object after the
+            commit.
+        change_list: list(dict). A list of changes introduced in this commit.
+    """
+    old_exp_version = exploration.version - 1
+    new_exp_version = exploration.version
+    exploration_stats = get_exploration_stats_by_id(
+        exploration.id, old_exp_version)
+    if exploration_stats is None:
+        handle_stats_creation_for_new_exploration(
+            exploration.id, exploration.version, exploration.states)
+        return
+
+    # This mapping from new state names to old ones accounts for circular
+    # renames and multiple renames within a commit. We will use this mapping
+    # to handle state renames in the statistics model.
+    new_to_old_state_names = exploration.get_state_names_mapping(change_list)
+    old_state_names = new_to_old_state_names.values()
+
+    # Handling state additions, deletions and renames.
+    for change_dict in change_list:
+        if change_dict['cmd'] == exp_domain.CMD_ADD_STATE:
+            if change_dict['state_name'] in old_state_names:
```

Comment:
I think the flow here is a bit suspect. Probably should start with a dict mapping old state names to their corresponding state stats. Then, walk through commit list, and update that dict as you go along so that the state stats are always correct, and at the end you'll have the desired new version of that dict (with the new state names as the keys). At the moment it's not clear to me that all cases are covered and that state identity is preserved.

Reply:
@seanlip If we are gonna take each commit and work on it as we go, we duplicate code that we wrote for the classifier framework. I'm just trying to reuse some portions so we can make it easier here. But, you're right, I'll see how I can make this flow more intuitive, it's kind of a jumble now.

Reply:
Makes sense -- but, let's get it right and intuitive here, and then later think about how to reduce the duplication. A mistake here can be costly.

Reply:
Done. The process looks much more cleaner now. PTAL